### PR TITLE
[ASCII-1224] Prevent deadlock in config.OnUpdate

### PR DESCRIPTION
**Pull request by dustmop**
### What does this PR do?

Since config.Set claims a mutex, and notificationReceivers currently is invoked within that mutex, any callback using config.Get would deadlock. Instead, clone the receives, then release the mutex before invoking them.

### Motivation

Users of config.OnUpdate must be allowed to call config.Get.

### Possible Drawbacks / Trade-offs

Allocates memory for each call to config.Set, assuming config.OnUpdate has been called at least once. It may be better to have two separate locks instead: one for Get/Set and another for the notificationReceivers list.

### Describe how to test/QA your changes

No behavioral change aside from preventing the deadlock. That fix is covered by a unit test.

### CODE REVIEW

File: pkg/config/model/viper.go [hunk 1]
```diff
@@ -89,13 +90,18 @@ func (c *safeConfig) Set(key string, value interface{}, source Source) {
...
-	c.configSources[source].Set(key, value)
-	c.mergeViperInstances(key)
+	// modify the config then release the lock to avoid deadlocks
+	var receivers []NotificationReceiver
+	func() {
+		c.Lock()
+		defer c.Unlock()
+		c.configSources[source].Set(key, value)
+		c.mergeViperInstances(key)
+		receivers = slices.Clone(c.notificationReceivers)
```
Comment (jeremy-hanna) [inline pkg/config/model/viper.go:100]: suggestion: I think this landed in go 1.21, should we use that? 

https://pkg.go.dev/slices#Clone

**Review by jeremy-hanna — APPROVED** [review]
LGTM 👍

File: pkg/config/model/viper.go [hunk 2]
```diff
@@ -89,13 +90,18 @@ func (c *safeConfig) Set(key string, value interface{}, source Source) {
...
+	}()
```
Comment (pgimalac) [inline pkg/config/model/viper.go:101]: **🥜 nitpick:** is the closure only used to be able to `defer` directly ? I feel like it makes the code less clear than just `Unlocking` 3 lines down, what do you think ?‏

File: pkg/config/model/viper.go [hunk 3]
```diff
@@ -89,13 +90,18 @@ func (c *safeConfig) Set(key string, value interface{}, source Source) {
...
 
 	// notifying all receiver about the updated setting
-	for _, receiver := range c.notificationReceivers {
+	for _, receiver := range receivers {
```
Comment (ogaca-dd) [inline pkg/config/model/viper.go:104]: You have a race issue as `receivers` is set line `100` by another goroutine
What do you think about the following code:
```go
var receivers []NotificationReceiver
c.Lock()
c.configSources[source].Set(key, value)
c.mergeViperInstances(key)
receivers = slices.Clone(c.notificationReceivers)
c.Unlock()

for _, receiver := range receivers {
...
```

File: pkg/config/model/viper.go (hunk 2 above)
Comment (dustmop) [reply pkg/config/model/viper.go:101]: Now that you mention it, I do think that's better, I will make that change.

File: pkg/config/model/viper.go (hunk 3 above)
Comment (dustmop) [reply pkg/config/model/viper.go:104]: Sure, that sounds good. It wasn't using a goroutine before, but regardless, I agree this looks better.

**Review by vickenty — APPROVED** [review]

**Review by gh123man — APPROVED** [review]

**Review by jennchenn — APPROVED** [review]

**Review by songy23 — APPROVED** [review]
Approval for OTel

**Review by ameske — APPROVED** [review]

**Review by mfpierre — APPROVED** [review]

**Review by dinooliva — APPROVED** [review]

**Review by spencergilbert — APPROVED** [review]

**Pull request merged**